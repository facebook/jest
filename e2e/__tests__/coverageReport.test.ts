/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import * as path from 'path';
import * as fs from 'graceful-fs';
import {wrap} from 'jest-snapshot-serializer-raw';
import {runYarnInstall} from '../Utils';
import runJest from '../runJest';

const DIR = path.resolve(__dirname, '../coverage-report');

beforeAll(() => {
  runYarnInstall(DIR);
});

test('outputs coverage report', () => {
  const {stdout, exitCode} = runJest(DIR, ['--no-cache', '--coverage'], {
    stripAnsi: true,
  });
  const coverageDir = path.join(DIR, 'coverage');

  // - the `setup.js` file is ignored and should not be in the coverage report.
  // - `SumDependency.js` is mocked and the real module is never required but
  //  is listed with 0 % coverage.
  // - `notRequiredInTestSuite.js` is not required but it is listed
  //  with 0 % coverage.
  expect(wrap(stdout)).toMatchSnapshot();

  expect(() => fs.accessSync(coverageDir, fs.constants.F_OK)).not.toThrow();
  expect(exitCode).toBe(0);
});

test('collects coverage only from specified file', () => {
  const {stdout} = runJest(
    DIR,
    [
      '--no-cache',
      '--coverage',
      '--collectCoverageFrom', // overwrites the one in package.json
      'file.js',
    ],
    {stripAnsi: true},
  );

  // Coverage report should only have `file.js` coverage info
  expect(wrap(stdout)).toMatchSnapshot();
});

test('collects coverage only from multiple specified files', () => {
  const {stdout} = runJest(
    DIR,
    [
      '--no-cache',
      '--coverage',
      '--collectCoverageFrom',
      'file.js',
      '--collectCoverageFrom',
      'otherFile.js',
    ],
    {stripAnsi: true},
  );

  expect(wrap(stdout)).toMatchSnapshot();
});

test('collects coverage only from specified files avoiding dependencies', () => {
  const {stdout} = runJest(
    DIR,
    [
      '--no-cache',
      '--coverage',
      '--collectCoverageOnlyFrom',
      'sum.js',
      '--',
      'sum.test.js',
    ],
    {stripAnsi: true},
  );

  // Coverage report should only have `sum.js` coverage info
  expect(wrap(stdout)).toMatchSnapshot();
});
