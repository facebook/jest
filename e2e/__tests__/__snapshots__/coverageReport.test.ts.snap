// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collects coverage only from multiple specified files 1`] = `
--------------|---------|----------|---------|---------|-------------------
File          | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
--------------|---------|----------|---------|---------|-------------------
All files     |     100 |      100 |     100 |     100 |                   
 file.js      |     100 |      100 |     100 |     100 |                   
 otherFile.js |     100 |      100 |     100 |     100 |                   
--------------|---------|----------|---------|---------|-------------------
`;

exports[`collects coverage only from specified file 1`] = `
----------|---------|----------|---------|---------|-------------------
File      | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
----------|---------|----------|---------|---------|-------------------
All files |     100 |      100 |     100 |     100 |                   
 file.js  |     100 |      100 |     100 |     100 |                   
----------|---------|----------|---------|---------|-------------------
`;

exports[`collects coverage only from specified files avoiding dependencies 1`] = `
----------|---------|----------|---------|---------|-------------------
File      | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
----------|---------|----------|---------|---------|-------------------
All files |    87.5 |      100 |      50 |     100 |                   
 sum.js   |    87.5 |      100 |      50 |     100 |                   
----------|---------|----------|---------|---------|-------------------
`;

exports[`outputs coverage report 1`] = `
-------------------------------------|---------|----------|---------|---------|-------------------
File                                 | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
-------------------------------------|---------|----------|---------|---------|-------------------
All files                            |      60 |        0 |      50 |      60 |                   
 coverage-report                     |   47.37 |        0 |      25 |      50 |                   
  file.js                            |     100 |      100 |     100 |     100 |                   
  notRequiredInTestSuite.js          |       0 |        0 |       0 |       0 | 8-19              
  otherFile.js                       |     100 |      100 |     100 |     100 |                   
  sum.js                             |    87.5 |      100 |      50 |     100 |                   
  sumDependency.js                   |       0 |        0 |       0 |       0 | 8-12              
 coverage-report/cached-duplicates/a |     100 |      100 |     100 |     100 |                   
  identical.js                       |     100 |      100 |     100 |     100 |                   
 coverage-report/cached-duplicates/b |     100 |      100 |     100 |     100 |                   
  identical.js                       |     100 |      100 |     100 |     100 |                   
-------------------------------------|---------|----------|---------|---------|-------------------
`;
